Dashboard redesign component

caneruguz opened the pull request:
## Purpose 
Step 1 towards building the dashboard.
This step includes basic architecture of a modular system called file-browser. It uses the new api2 and has multiple components to take care of view. 

## Changes
- Old dashboard parts are removed from view and code including onboarding and watchlist
- Project organizer is refactored to use a much more skinny version since direct editing and moving of collections is removed
- File browser module is added to be a wrapper for the entire project organizer view and includes sub components that cover distinct tasks 
    - Breadcrumbs : updates the current location of folder
   - Collections : list of collections that updates the project view
   - Filter List: A list of tags and users that updates the project view (not functional yet but view is there)
   - Information Panel: includes information pertinent to selected row.
- Double clicking on project opens in new folder. (currently using double click on placeholder text called "Open" until view is finalized. 

This PR is a first step so final functionalities are incomplete. Aspects to be checked are:
- The parts that are deleted are deleted properly
- No hanging unused code
- Tests are passing and tests that belong to removed code are removed
- The file-browser module is working: 
    - all subcomponents are displaying, 
    - information panel is working, 
    - collections and links are loading correct api end points
 
There is no QA needed for this process. 

Pertinent to these jira issues:
https://openscience.atlassian.net/browse/OSF-4795
https://openscience.atlassian.net/browse/OSF-4796
https://openscience.atlassian.net/browse/OSF-4798

lbanner commented:
JIRA ticket # @caneruguz ?

caneruguz commented:
@lbanner thanks for catch, I added them to the description. the issue tickets are in the last commit too so they will be shown in jira. 

I'll look into the travis error now.

samchrisinger commented on website/static/js/file-browser.js:
```diff
@@ -0,0 +1,284 @@
+/**
+ *  Options for fileBrowser
+ */
+var defaults = {
```
Are these used? 

Also it may be cleaner to put all FileBrowser specific values under the FileBrowser namespace.

caneruguz replied:
Good point, they are not actually used. Removing them now.

samchrisinger commented on website/static/js/file-browser.js:
```diff
@@ -0,0 +1,284 @@
+        self.showInfo = m.prop(false);
+
+        self.collections = [
+            { id:1, type : 'collection', label : 'All My Projects', path : 'users/me/nodes/', pathOptions : {  query : { 'filter[registration]' : 'false'} } },
```
Will `pathOptions` ever be anything but query string key/value pairs? There may unnecessary nesting here.

caneruguz replied:
Changed as suggested

samchrisinger commented on website/static/js/file-browser.js:
```diff
@@ -0,0 +1,284 @@
+        self.filesData = m.prop($osf.apiV2Url(self.collections[0].path, self.collections[0].pathOptions));
+
+        self.breadcrumbs = m.prop([
+            { label : 'All My Projects', url : 'http://localhost:8000/v2/users/me/nodes/?filter%5Bregistration%5D=false', type : 'collection'}
```
Please use `encodeURIComponent` or some other helper method for encoding query string params.

samchrisinger commented on website/static/js/file-browser.js:
```diff
@@ -0,0 +1,284 @@
+            { tag : 'something', type : 'filter'}
+        ];
+
+        self.updateFilesData = function(linkObject) {
```
Consider moving `self.<name>` methods to the prototype

samchrisinger commented on website/static/js/file-browser.js:
```diff
@@ -0,0 +1,284 @@
+        self.updateList = function(element, isInit, context){
+            if(!self.isLoadedUrl) {
+                var el = element || document.getElementById('pOrganizer');
+                m.mount(el, m.component( ProjectOrganizer, { filesData : self.filesData, updateSelected : self.updateSelected, updateFilesData : self.updateFilesData}));
```
Just my opinion, but the indentation here is not very readable.

caneruguz replied:
Changed as suggested

samchrisinger commented on website/static/js/file-browser.js:
```diff
@@ -0,0 +1,284 @@
+        self.activeCollection = m.prop(1);
+        self.updateCollection = function(coll) {
+            self.activeCollection(coll.id);
+            var linkObject = {
```
It would be good to formalize the structure of `linkObject` as a JS class, e.g.:

```javascript 
var LinkObject = function(type, data, label, id) {
    ...
};
```

caneruguz replied:
Changed as suggested

samchrisinger commented on website/static/js/file-browser.js:
```diff
@@ -0,0 +1,284 @@
+        self.updateSelected = function(selectedList){
+            self.selected(selectedList);
+            console.log(self.selected());
+        }.bind(self);
```
This bind is probably not needed (`self` looks to be included already via closure)

caneruguz replied:
Changed as suggested

samchrisinger commented on website/static/js/file-browser.js:
```diff
@@ -0,0 +1,284 @@
+        // Refresh the Grid
+        self.updateList = function(element, isInit, context){
+            if(!self.isLoadedUrl) {
+                var el = element || document.getElementById('pOrganizer');
```
Is `element` ever variable? Can it ever be undefined?

Falling back to a hard-coded id seems a bit dangerous (although hopefully we never have more than one PO on a page)

caneruguz replied:
Good catch, the element doesn't even refer to poOrganizer. I added a better method of finding, though that fallback should not be needed.

samchrisinger commented on website/static/js/file-browser.js:
```diff
@@ -0,0 +1,284 @@
+        // BREADCRUMBS
+        self.updateBreadcrumbs = function(linkObject){
+            var crumb = {
+                label : linkObject.label,
```
Like with `LinkObject`, if the code requires these objects to have a certain format, it may be beneficial to make that explicit as a class.

caneruguz replied:
Good point, turns out a lot of the models I use need a class structure. I made the changes to include this one and all others

samchrisinger commented on website/static/js/file-browser.js:
```diff
@@ -0,0 +1,284 @@
+                return;
+            }
+            if (linkObject.type === 'breadcrumb'){
+                self.breadcrumbs().splice(linkObject.index+1, self.breadcrumbs().length-linkObject.index-1);
```
This line doesn't appear to do anything.

samchrisinger replied:
Sorry, splice !== slice

samchrisinger replied:
I may be missing it, but can you please point me to where the `linkObjects`'s `index` is set?

caneruguz replied:
Sorry my bad, I updated the linkObject class to reflect this better. However I thought I was using the right method, splice is what I am trying to do.

samchrisinger commented on website/static/js/file-browser.js:
```diff
@@ -0,0 +1,284 @@
+            if (linkObject.type === 'user') {
+                return $osf.apiV2Url('users/' + linkObject.id + '/nodes', {});
+            }
+            return $osf.apiV2Url('nodes/' + linkObject.data.uid + '/children', {}); // If type === node
```
It might be better to just do an explicit `else if`, and if you're feeling really nice add an error raising `else` branch. (Also the `if`s above could be `else if`s

caneruguz replied:
I was using jslint suggestion to not use elses if returns are involved but from a readibility it makes more sense to specift the logic. changed as suggested.

samchrisinger commented on website/static/js/file-browser.js:
```diff
@@ -0,0 +1,284 @@
+ */
+var Collections  = {
+    controller : function (args) {
+        var self = this;
```
A name more specific than `args` would improve readability

caneruguz replied:
I'm struggling with this, it's something mithril uses and it's supposed to be super general.

samchrisinger replied:
I'm guessing `args` is an instance of one of your controllers?

samchrisinger commented on website/static/js/file-browser.js:
```diff
@@ -0,0 +1,284 @@
+        return m('.fb-collections', m('ul', [
+            args.list.map(function(item, index, array){
+                selectedCSS = item.id === args.activeCollection() ? '.active' : '';
+                return m('li' + selectedCSS,
```
Might be better to use `className` attr rather than string concatenation

caneruguz replied:
Changed as suggested

samchrisinger commented on website/static/js/file-browser.js:
```diff
@@ -0,0 +1,284 @@
+ */
+var Information = {
+    view : function (ctrl, args) {
+        var template = '';
```
Again, it looks like `args` is expected to be an instance of some class and could probably be named better.

samchrisinger commented on website/static/js/pages/dashboard-page.js:
```diff
@@ -10,29 +9,15 @@ var ko = require('knockout');
-var ProjectOrganizer = require('js/projectorganizer').ProjectOrganizer;
+var m = require('mithril'); // exposes mithril methods, useful for redraw etc.
 
 var url = '/api/v1/dashboard/get_nodes/';
```
Can we do this with a V2 route?

caneruguz replied:
This is not needed anymore so I removed

samchrisinger commented on website/static/js/pages/dashboard-page.js:
```diff
@@ -10,29 +9,15 @@ var ko = require('knockout');
-    $osf.applyBindings({nodes: registrationSelection, enableComponents: true}, '#obRegisterProject');
-    $osf.applyBindings({nodes: uploadSelection}, '#obUploader');
-    $osf.applyBindings({nodes: allNodes}, '#obCreateProject');
+    //.. Getting nodes
```
Oh. Is this request needed?

caneruguz replied:
No, I thought at some point I might need that call but since the switch to V2 that's not necessary

samchrisinger commented on website/static/js/pages/dashboard-page.js:
```diff
@@ -40,6 +25,8 @@ request.fail(function(xhr, textStatus, error) {
     });
 });
 
+var nodesListUrl = '/api/v1/dashboard/get_nodes/';
```
This seems unused

caneruguz replied:
Yes, these old calls are unused, I'll remove it.

samchrisinger commented on website/static/js/project-organizer.js:
```diff
@@ -0,0 +1,488 @@
+    },
+    queryTokenizer: Bloodhound.tokenizers.whitespace,
+    remote: {
+        url: '/api/v1/search/projects/?term=%QUERY&maxResults=20&includePublic=yes&includeContributed=no',
```
Consider using a helper to build this url. `$.param` is one option

samchrisinger commented on website/static/js/project-organizer.js:
```diff
@@ -0,0 +1,488 @@
+/**
```
I see this is just a renaming of the old file.

Feel free to ignore my comments

samchrisinger commented on website/templates/dashboard.mako:
```diff
@@ -1,69 +1,22 @@
 <%inherit file="base.mako"/>
 <%def name="title()">Dashboard</%def>
 
+<%def name="container_class()">container-xxl</%def>
```
Is this used?

caneruguz replied:
Yes, it allows certain pages on osf to go up to 1600px wide like the Wiki. This page needs it too.

samchrisinger commented:
@caneruguz left you some questions, and am in general curious about what your remaining TODOs are. 

Looks good so far though.

felliott commented:
In `website/static/js/project-organizer.js` there are three functions I can't find referenced anywhere else: `_gotoEvent`, ` addFormKeyBindings`, and `triggerClickOnItem`.  Do these need to be deleted?

felliott commented:
How does treebeard handle errors when the requested resource doesn't exist?  I don't have a 'Caner Uguz' (id: 8q63f) user in my dev OSF, so when I click on the user filter my window no longer responds.  Instead I get the following error message in my Console:

```
SyntaxError: JSON.parse: unexpected character at line 1 column 1 of the JSON data
```
from the url `http://localhost:8000/v2/users/8q63f/nodes/?related_counts=true&_=1446844068568`.  The 404 response body is: 

````
{ "errors": [
  {"detail": "Not found"}
] }
````

felliott commented:
Hey @caneruguz,

The new organizer looks really, *really*, **really** nice!  The code looks pretty good to me, too. I'm going to move all three tickets back to WIP, with the following recommended actions for each ticket:

* [OSF-4795 - Removing unused code](https://openscience.atlassian.net/browse/OSF-4795) - I noted some files and code that may be able to be removed, but I defer to your expertise on this.  Once you've looked at my comments and decided whether they're valid, please move this to Ready-To-Merge.

* [OSF-4796 - New js file for the project organizer](https://openscience.atlassian.net/browse/OSF-4796) - The trailing slashes and the error handling are the blockers here.  I'm not sure if error-handling is in the scope of this ticket, or if it's something you're planning on tackling down the road.  If error handling comes later, you can move this to Ready-To-Merge after fixing the slashes, but please leave a comment here or in the JIRA master ticket pointing to a ticket for error-handling.  If you decide to fix it now, move this back to CR2 when you're done, and assign me.

* [OSF-4798 - Breadcrumbs](https://openscience.atlassian.net/browse/OSF-4798) - Once you have nested breadcrumb parents working, move this to CR2, assign me, and I'll review.

Once more I want to stress that this looks fantastic, and that this code was a real pleasure to read!  I felt I learned a lot from seeing how you approached this.  I'm excited to use this when I tackle the manyfile uploader.  Thanks again, @caneruguz !

Cheers,
Fitz

caneruguz commented:
Thanks @felliott . This is an amazingly well done review with great eye for detail, especially with something as important as removing unused code. I will come back to you if I have questions, otherwise I'll move it back to CR2 when done. Cheers!

caneruguz commented:
@felliott for this comment: https://github.com/CenterForOpenScience/osf.io/pull/4526#issuecomment-154542322

Treebeard gives an error when a link is not provided but not an error when the returned value is wrong. This could probably be improved but those links are not going to be hard coded when the api finishes so we shold revisit it then.

caneruguz commented:
@felliott I made changes as requested. Can you review the commits and let me know if the implementations are good?

felliott commented:
Everything looks good to me!  I'm moving this to RTM.

sloria commented:
There should be some error handling if the API server cannot be reached or returns an error response. Currently, when i open the dashboard without running the API server, the project organizer renders without projects, and no error feedback is given.

sloria commented:
As per COSDev, use `lowerCamelCase` for reusable JS modules, e.g. `fileBrowser.js` rather than `file-browser.js`.

Also, a more appropriate name for the dashboard component is `Dashboard` rather than `FileBrowser`.

sloria commented on website/static/js/file-browser.js:
```diff
@@ -0,0 +1,290 @@
+        self.updateList = function(element, isInit, context){
+            if(!self.isLoadedUrl) {
+                var el = element || $(self.wrapperSelector).find('.fb-main').get(0);
+                m.mount(el,
```
Can you explain why it's necessary to call `mount` here rather than using `m.component` in the view?

sloria commented:
I have a few minor comments and question above. None are blockers to mere. Considering this is still in prototype stage, this is good to merge.

caneruguz commented:
@sloria thanks for the review. I made changes in the second PR for the project organizer work, it should come your way very soon (I'll link it here when I submit it). 
- I changed the file names
- I added error handling in case api server returned with an error
- I used component in view rather than mount as you noted above. 

The one thing I did not change is the naming of FileBrowser to Dashboard. The file browser tool is supposed to be reusable and may be used in places outside the dashboard in its entirety, for instance it may eventually replace all File views. We still have time to make this name change depending on where things go. I would suggest to revisit it before final merge.
